src/basilisp/contrib/nrepl_server.lpy

```diff
@@ -14,18 +15,21 @@
   "The logger for this namespace."
   (logging/getLogger (namespace ::)))
 
+(defmacro ^:private trace [& values]
+  `(when (.isEnabledFor logger basilisp.logconfig/TRACE)
+     (.log logger basilisp.logconfig/TRACE (str/join " " [~@values]))))
 (defmacro ^:private debug [& values]
-  (when (.isEnabledFor logger logging/DEBUG)
-    `(.debug logger (str/join " " [~@values]))))
+  `(when (.isEnabledFor logger logging/DEBUG)
+    (.debug logger (str/join " " [~@values]))))
```

Comment:
I think these changes to move the `(.isEnabledFor logger ...)` checks into the macro expansion are probably not desirable. Calling `(.debug logger ...)` (and so on for every selected log level) [performs this check already](https://github.com/python/cpython/blob/3.12/Lib/logging/__init__.py#L1517-L1527).

I'd be either in favor of leaving the `(.isEnabledFor logger ...)` checks outside of the macroexpansion (so the calls are completely elided if the log level isn't enabled) or removing them entirely and delegating to the `logging.Logger` runtime check.

Reply:
I wanted to set the log level dynamically, which is why I moved it inside the macro expansion. This approach is useful when the log level is set in a start script or a user.lpy repl helper. I understand that your concerns aren't about preferring static logging setup over dynamic, but rather about the potential additional cost caused by checking the log level more frequently than necessary.

I can't delegate directly to the logger with a single `(.debug ...)` in the macro expansion because that would mean evaluating the arguments passed to the function regardless of whether they are logged or not, which I want to avoid. This is why I added the `when` check.

I'll remove the redundant log level checks and hopefully can continue the discussion from the improved version.

Thanks!

Reply:
Updated.

Just to recap, my goal is to respect the logging level when it's set dynamically, while avoiding the evaluation and concatenation of the `values` arguments if the level is not set.

```clojure
(defmacro ^:private debug [& values]
  `(when (.isEnabledFor logger logging/DEBUG)
    (.log logger logging/DEBUG (str/join " " [~@values]))))
```

Reply:
> [...] I understand that your concerns aren't about preferring static logging setup over dynamic, but rather about the potential additional cost caused by checking the log level more frequently than necessary.
> 
> I can't delegate directly to the logger with a single (.debug ...) in the macro expansion because that would mean evaluating the arguments passed to the function regardless of whether they are logged or not, which I want to avoid. This is why I added the when check.

To be honest, I'm not and never have been concerned about the cost of evaluating arguments to logging functions. It's one of the most overblown micro-optimizations in the software engineering realm (you can see I often use f-strings in logging within Basilisp Python code and have disabled the PyLint check against this because I just don't agree with the concern here).

Unless the logging statement is making a call to a database or other networked service or constantly evaluating arguments in a tight loop, I don't think there's much of an argument to be made for the performance cost of such calls.

My main concern was duplicating logic already being handled by the logger, but since you've moved to using `(.log logger ...)` I'm no longer concerned about it.

I think what you have now is fine, although again I'm not really concerned about performance.

Reply:
Great thanks, that was a very good catch!